artifact/src/version.rs

```diff
@@ -0,0 +1,243 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::{borrow::Cow, fmt, str::FromStr};
+
+use daft::Diffable;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+/// An artifact version.
+///
+/// This is a freeform identifier with some basic validation. It may be the
+/// serialized form of a semver version, or a custom identifier that uses the
+/// same character set as a semver.
+///
+/// For the exact regex pattern accepted by [`ArtifactVersion::new`], see
+/// [`ArtifactVersion::REGEX`].
+///
+/// # Ord implementation
+///
+/// `ArtifactVersion`s are not intended to be sorted, just compared for
+/// equality. `ArtifactVersion` implements `Ord` only for storage within sorted
+/// collections.
+#[derive(Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord, Diffable)]
+#[cfg_attr(any(test, feature = "proptest"), derive(test_strategy::Arbitrary))]
+#[cfg_attr(any(test, feature = "schemars"), derive(schemars::JsonSchema))]
+#[cfg_attr(any(test, feature = "schemars"), schemars(regex = Self::REGEX))]
+#[daft(leaf)]
+pub struct ArtifactVersion(
+    #[cfg_attr(any(test, feature = "proptest"), strategy(PROPTEST_REGEX))]
+    #[cfg_attr(any(test, feature = "proptest"), map(Cow::Owned))]
+    #[cfg_attr(
+        any(test, feature = "schemars"),
+        schemars(regex = "Self::REGEX")
+    )]
+    Cow<'static, str>,
+);
+
+impl ArtifactVersion {
+    /// The maximum length of a version string.
+    ///
+    /// This matches the length allowed in Omicron database storage.
+    pub const MAX_LEN: usize = 63;
+
+    /// A regular expression that matches a valid version string.
+    ///
+    /// This is the set of characters allowed in a semver, though without any
+    /// additional structure. We expect non-semver identifiers to only use these
+    /// characters as well.
+    pub const REGEX: &str = r"^[a-zA-Z0-9._+-]{1,63}$";
```

Comment:
`_` is not allowed in semver, although I have no concern with allowing it.

Reply:
Ah good catch!